# landfill-microcosms
16s amplicon data from lab-built microcosms 

# Study description

The study used simulated landfill microcosms and tracked microbial populations over a 65 day period. Fifteen microcosms were set up identically on day 0, and DNA was sampled from all 15 microcosms on days 6 (T1), 20 (T2), 29 (T3), and 40 (T4). On day 40, three microcosms were destructively sampled for chemical measurements including sulfide, sulfate, total iron, and ferrous iron. On day 44, triplicate microcosms were "spiked" with different additives (sodium sulfate, antibiotics, and Fe(OH)3). Two additional DNA samples were taken at 50 (T5) days and 65 (T6) days of incubation from the remaining 12 microcosms. At the end of the 65 day period, all microcosms were destructively sampled for chemical concentrations of sulfide, sulfate, total iron, and ferrous iron. 



# Tutorials and papers used to produce these markdown files: 

Callahan et al. (2016). Bioconductor Workflow for Microbiome Data Analysis: from raw reads to community analyses [Version 2]. F1000Reasearch. 

Handley et al., "2018 ASM Using R to Analyze the Bacterial Microbiome Workshop"

https://benjjneb.github.io/dada2/tutorial_1_8.html

https://astrobiomike.github.io/

http://joey711.github.io/phyloseq-demo/phyloseq-demo.html

https://mibwurrepo.github.io/Microbial-bioinformatics-introductory-course-Material-2018/citation.html


Further reading: 

http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html

Random Forest in R example: https://rpubs.com/michberr/randomforestmicrobe

LEfSe: https://bitbucket.org/biobakery/biobakery/wiki/lefse

ANCOM: https://www.ncbi.nlm.nih.gov/pubmed/26028277
